packages/app-builder-lib/src/node-module-collector/npmNodeModulesCollector.ts

```diff
@@ -15,53 +14,34 @@ export class NpmNodeModulesCollector extends NodeModulesCollector<NpmDependency,
     return ["list", "-a", "--include", "prod", "--include", "optional", "--omit", "dev", "--json", "--long", "--silent"]
   }
 
-  protected extractRelevantData(npmTree: NpmDependency): NpmDependency {
-    const tree = super.extractRelevantData(npmTree)
-    const { optionalDependencies, _dependencies } = npmTree
-    return { ...tree, optionalDependencies, _dependencies }
-  }
-
-  protected extractProductionDependencyTree(tree: NpmDependency): DependencyTree {
-    const _deps = tree._dependencies ?? {}
-
-    let deps = tree.dependencies ?? {}
-    let implicitDependenciesInjected = false
-
-    if (Object.keys(_deps).length > 0 && Object.keys(deps).length === 0) {
-      log.debug({ name: tree.name, version: tree.version }, "injecting implicit _dependencies")
-      deps = this.allDependencies.get(`${tree.name}@${tree.version}`)?.dependencies ?? {}
-      implicitDependenciesInjected = true
-    }
-
-    const dependencies = Object.entries(deps).reduce<DependencyTree["dependencies"]>((acc, curr) => {
-      const [packageName, dependency] = curr
-      if (!_deps[packageName] || Object.keys(dependency).length === 0) {
-        return acc
+  protected collectAllDependencies(tree: NpmDependency) {
+    for (const [key, value] of Object.entries(tree.dependencies || {})) {
+      const { _dependencies = {}, dependencies = {} } = value
+      const isDuplicateDep = Object.keys(_dependencies).length > 0 && Object.keys(dependencies).length === 0
+      if (isDuplicateDep) {
+        continue
       }
-      if (implicitDependenciesInjected) {
-        const { name, version, path, workspaces } = dependency
-        const simplifiedTree: ParsedDependencyTree = { name, version, path, workspaces }
-        return {
-          ...acc,
-          [packageName]: { ...simplifiedTree, implicitDependenciesInjected },
-        }
-      }
-      return {
-        ...acc,
-        [packageName]: this.extractProductionDependencyTree(dependency),
-      }
-    }, {})
-
-    const { name, version, path: packagePath, workspaces } = tree
-    const depTree: DependencyTree = {
-      name,
-      version,
-      path: packagePath,
-      workspaces,
-      dependencies,
-      implicitDependenciesInjected,
+      this.allDependencies.set(`${key}@${value.version}`, value)
+      this.collectAllDependencies(value)
     }
-    return depTree
+  }
+
+  protected extractProductionDependencyGraph(tree: NpmDependency, dependencyId: string): void {
+    if (this.productionGraph[dependencyId]) return
+
+    const { _dependencies: prodDependencies = {}, dependencies = {} } = tree
+    const isDuplicateDep = Object.keys(prodDependencies).length > 0 && Object.keys(dependencies).length === 0
+    const resolvedDeps = isDuplicateDep ? (this.allDependencies.get(dependencyId)?.dependencies ?? {}) : dependencies
+    // don't delete this, it's used to prevent infinite loops
+    this.productionGraph[dependencyId] = { dependencies: [] }
```

Comment:
Can you elaborate on this a bit more? This comment makes me feel like there's an alternative way to write this.

Reply:
Yes. 

 // Initialize with empty dependencies array first to mark this dependency as "in progress"
    // This prevents infinite loops when circular dependencies are encountered

Reply:
After this initialization, if there are libraries with the same name+version later, they will not be searched recursively again, which can prevent infinite loops.